### PR TITLE
Add Gather PDF Authors auxiliary module

### LINKED ISSUE #6309: A module that collects PDF metadata
As a pentester, sometimes I (well not me, our pentesters) run into an external site that has a collection of white papers in PDF format. It would be nice to collect the author names as a way to enumerate employee names.

### PR DESCRIPTION
**Pull request by bcoles**
This module takes a list of PDF URLs, downloads the PDFs, and extracts the author's name from the document metadata.

Fix #6309

This module requires the `pdf-reader` gem as a dependency. The gem will need to be added to the gemspec.

## Verification

- [x] Start `msfconsole`
- [x] `use auxiliary/gather/http_pdf_authors`
- [x] `set URL_LIST [FILE]`
- [x] **Verify** the module retrieves the URLs and extracts the author's name from each document

### Example Output

```
[*] Processing 8 URLs...
[*] Downloading 'http://127.0.0.1:80/test.pdf'
[*] HTTP 200 -- Downloaded PDF (89283 bytes)
[*]  12.50% done (1/8 files)
[*] Downloading 'http://127.0.0.1/test2.pdf'
[*] HTTP 200 -- Downloaded PDF (636661 bytes)
[+] PDF Author: sqlmap developers
[*]  25.00% done (2/8 files)
[*] Downloading 'http://127.0.0.1/test3.pdf'
[*] HTTP 200 -- Downloaded PDF (167478 bytes)
[+] PDF Author: Evil1
[*]  37.50% done (3/8 files)
[*] Downloading 'http://127.0.0.1/test4.pdf'
[*] HTTP 200 -- Downloaded PDF (38867 bytes)
[+] PDF Author: Administrator
[*]  50.00% done (4/8 files)
[*] Downloading 'http://127.0.0.1/test5.pdf'
[*] HTTP 200 -- Downloaded PDF (34312 bytes)
[+] PDF Author: ekama
[*]  62.50% done (5/8 files)
[*] Downloading 'http://127.0.0.1/doesnotexist.pdf'
[*] HTTP 404 -- Downloaded PDF (289 bytes)
[-] Could not parse PDF: PDF is malformed
[*]  75.00% done (6/8 files)
[*] Downloading 'https://127.0.0.1/test.pdf'
[-] Connection failed: Failed to open TCP connection to 127.0.0.1:443 (Connection refused - connect(2) for "127.0.0.1" port 443)
[*] Downloading 'https://127.0.0.1:80/test.pdf'
[-] Connection failed: SSL_connect returned=1 errno=0 state=unknown state: unknown protocol

[+] Found 4 authors: sqlmap developers, Evil1, Administrator, ekama
[*] Auxiliary module execution completed
```

**With OUTFILE**

```
msf auxiliary(http_pdf_authors) > set outfile /root/output
outfile => /root/output
msf auxiliary(http_pdf_authors) > run

[*] Processing 8 URLs...
[*] Downloading 'http://127.0.0.1:80/test.pdf'
[*] HTTP 200 -- Downloaded PDF (89283 bytes)
[*]  12.50% done (1/8 files)
[*] Downloading 'http://127.0.0.1/test2.pdf'
[*] HTTP 200 -- Downloaded PDF (636661 bytes)
[+] PDF Author: sqlmap developers
[*]  25.00% done (2/8 files)
[*] Downloading 'http://127.0.0.1/test3.pdf'
[*] HTTP 200 -- Downloaded PDF (167478 bytes)
[+] PDF Author: Evil1
[*]  37.50% done (3/8 files)
[*] Downloading 'http://127.0.0.1/test4.pdf'
[*] HTTP 200 -- Downloaded PDF (38867 bytes)
[+] PDF Author: Administrator
[*]  50.00% done (4/8 files)
[*] Downloading 'http://127.0.0.1/test5.pdf'
[*] HTTP 200 -- Downloaded PDF (34312 bytes)
[+] PDF Author: ekama
[*]  62.50% done (5/8 files)
[*] Downloading 'http://127.0.0.1/doesnotexist.pdf'
[*] HTTP 404 -- Downloaded PDF (289 bytes)
[-] Could not parse PDF: PDF is malformed
[*]  75.00% done (6/8 files)
[*] Downloading 'https://127.0.0.1/test.pdf'
[-] Connection failed: Failed to open TCP connection to 127.0.0.1:443 (Connection refused - connect(2) for "127.0.0.1" port 443)
[*] Downloading 'https://127.0.0.1:80/test.pdf'
[-] Connection failed: SSL_connect returned=1 errno=0 state=unknown state: unknown protocol

[+] Found 4 authors: sqlmap developers, Evil1, Administrator, ekama
[*] Writing data to /root/output...
[*] Auxiliary module execution completed
msf auxiliary(http_pdf_authors) > cat /root/output
[*] exec: cat /root/output

sqlmap developers
Evil1
Administrator
ekamamsf auxiliary(http_pdf_authors) > 
```

### CODE REVIEW

**Comment by bcoles** [comment]
As expected, Travis failed for some builds due to missing `pdf-reader` dependency.

I refuse to write a PDF parser.

Are third-party dependencies acceptable for a single module?

What's best practice for adding a dependency? Should it be added as a `spec.add_runtime_dependency` in the `metasploit-framework.gemspec` or added to `Gemfile.lock` ?

**Comment by busterb** [comment]
add it to megasploit-framework.gemspec, then do 'bundle' and commit the lines that get added to Gemfile.lock by bundler for this dependency

one of the goal of external modules is to make this no longer a global dependency, but we're not quite there yet. be sure to include a note of what depends on this in metasploit-framework.gemspec though.

**Comment by bcoles** [comment]
@busterb cool, thanks. I've added the dep.

**Comment by dmohanty-r7** [comment]
Works for me:

```
msf > use auxiliary/gather/http_pdf_authors
msf auxiliary(http_pdf_authors) > set URL_LIST <path>/pdf_file_list
URL_LIST => <path>/pdf_file_list
msf auxiliary(http_pdf_authors) > run

[*] Processing 1 URLs...
[*] Downloading 'http://127.0.0.1:8000/building-payloads-tutorial.pdf'
[*] HTTP 200 -- Downloaded PDF (507439 bytes)
[+] PDF Author: Rapid7
[*] 100.00% done (1/1 files)

[+] Found 1 authors: Rapid7
[*] Auxiliary module execution completed
```

**Comment by dmohanty-r7** [comment]
One item of discussion I wanted to bring up is why point to a file that has the list of pdf addresses as opposed to just providing a list of pdf addresses?

**Comment by bcoles** [comment]
@dmohanty-r7 It seemed like a good idea at the time. Personal preference.

Mostly because I saw writing this module as a quick win to close off an issue that's been open since 2015.

Some thoughts:

How would the user supply the URLs as an alternative to using a file? Comma separated or space separated in a `OptString` ? This would be error prone comparable to new-line separated format which naturally lends itself to file read for processing.

Additionally, there could be many URLs and each URL could be quite long. Attempting to paste a few KB of URLs into a `set` command in `msfconsole` is not my preferred user experience.

On the other hand, if you wanted to run the module against a single PDF, having to write the PDF URL to a file is cumbersome.

**Comment by bcoles** [comment]
The original use case was to extract the author names from a collection of documents on a website. The typical workflow would involve extracting PDF links from HTML, perhaps by grepping a directory listing, or sitemap, or google `filetype:pdf` search results.

An alternative approach would be automating the process by removing the requirement for a list of PDF URLs and instead allowing the user to specify a URL, then scraping the URL for all PDF links. This introduces issues of URL scope but could be useful for directory listings.

We could do both approaches as two separate module actions. One to parse a list of pre-defined PDF URLs for analysis, the other to parse a list of URLs to scrape for PDF URLs for subsequent analysis.

On an unrelated topic, on the topic of collecting PDFs from one or more hosts, limiting the module to run against a specified `rhost` as per the original use case seemed like an unnecessary restriction so I opted to allow arbitrary URLs.

Admittedly, limiting by `rhost` would have allowed using the `HttpClient` mixin which is cleaner than instantiating `Net::HTTP` directly and manually implementing the `PROXY` related options. Still, I feel allowing arbitrary URLs is more user friendly.

**Comment by bcoles** [comment]
I've added an option allowing users to specify a single URL with `URL`. If both a `URL` and `URL_LIST` are specified, the module will favor the `URL`.

Also added documentation.

File: modules/auxiliary/gather/http_pdf_authors.rb [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+      'Description' => %q{
+        This module downloads PDF files and extracts the author's
+        name from the document metadata.
+      },
+      'License'     => MSF_LICENSE,
+      'Author'      => 'Brendan Coles <bcoles[at]gmail.com>'))
+    register_options(
+      [
+        OptString.new('URL', [ false, 'The URL of a PDF to analyse', '' ]),
+        OptString.new('URL_LIST', [ false, 'File containing a list of PDF URLs to analyze', '' ]),
```
Comment (wvu-r7) [inline modules/auxiliary/gather/http_pdf_authors.rb]: This should be `OptPath`. FWIW, `OptString` can take `file:`.

Comment (bcoles) [inline modules/auxiliary/gather/http_pdf_authors.rb]: Yeah it was `OptPath` but I changed it to `OptString`.

The reason for this is that `OptPath` defaults to the working directory, not `nil` or `''`, so later the check fails when I use:

```ruby
    return [ datastore['URL'] ] unless datastore['URL'].to_s.eql? ''

    if datastore['URL_LIST'].to_s.eql? ''
      fail_with Failure::BadConfig, 'No URL(s) specified'
    end
```

Comment (bcoles) [inline modules/auxiliary/gather/http_pdf_authors.rb]: Also I invoke the "but other modules do it too" defense:

```
# grep -rn OUTFILE modules/auxiliary/ | grep OptString
modules/auxiliary/scanner/http/enum_wayback.rb:24:        OptString.new('OUTFILE', [ false, "Where to output the list for use"])
modules/auxiliary/gather/search_email_collector.rb:27:        OptString.new('OUTFILE', [ false, "A filename to store the generated email list"]),
modules/auxiliary/gather/shodan_search.rb:44:        OptString.new('OUTFILE', [false, 'A filename to store the list of IPs']),
```

Comment (wvu-r7) [inline modules/auxiliary/gather/http_pdf_authors.rb]: Honestly, I think `file:` is a bit of tribal knowledge not everyone will know about. I think what you've decided upon is fine.

Comment (dmohanty-r7) [inline modules/auxiliary/gather/http_pdf_authors.rb]: @bcoles As someone who was not aware of the `file:`, the URL option is much appreciated, as is the documentation. Well done!

**Comment by bcoles** [comment]
Thanks, but I was literally just about to add the scraping :V

**Comment by wvu-r7** [comment]
Well, you're a committer, aren't you? ;)

**Comment by bcoles** [comment]
Not yet. All good though. I'm too tired to write a decent description. The scraping works, but he module description and documentation are as clear as mud.

**Comment by dmohanty-r7** [comment]
## Release Notes
This module downloads PDF files and extracts the author's name from the document metadata.

**Pull request merged**